### PR TITLE
Mqtt managed client

### PR DESCRIPTION
**Pull request by skotl**
Migrate to managed MQTT client for more resilience

Note that this has a dependency on #662 

This is not a breaking change on its own but does incorporate breaking changes from #662

## Breaking change
Yes, via #662

## Proposed change
Move to MqttNet.Extensions.ManagedClient away from the simplified client
Wrap this up in an assurer class that the `MessageSender` class depends on for assuring a client is available (but is not necessarily connected).

## Type of change

- [x] Dependency upgrade - needs #662 
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration) - resilience
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/Extensions/NetDaemon.Extensions.MqttEntityManager/MessageSender.cs [hunk 1]
```diff
@@ -20,75 +18,44 @@ namespace NetDaemon.Extensions.MqttEntityManager;
...
+    ///     Publish a message to the given topic
     /// </summary>
     /// <param name="topic"></param>
     /// <param name="payload">Json structure of payload</param>
     /// <param name="retain"></param>
     public async Task SendMessageAsync(string topic, string payload, bool retain = false)
     {
-        using var mqttClient = _mqttFactory.CreateMqttClient();
-        await ConnectAsync(mqttClient);
+        var mqttClient = _assuredMqttConnection.GetClientOrThrow();
```
Comment (helto4real) [inline src/Extensions/NetDaemon.Extensions.MqttEntityManager/MessageSender.cs:42]: This should be async. Then you could await it with a CancellationToken that are time out for example. Now you mix sync/async code and that often leads to unpredictable results like blocking. You could for example provide the CancellationToken to this method and that should be used while waiting for connection.

File: src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        _connectionTask = Task.Run(() => ConnectAsync(mqttConfig.Value, mqttFactory));
+    }
+
+    /// <summary>
+    /// Ensures that an MQTT client is available, retrying if necessary, and throws if connection is impossible
+    /// </summary>
+    /// <returns></returns>
+    /// <exception cref="MqttConnectionException">Timed out while waiting for connection</exception>
+    public IManagedMqttClient GetClientOrThrow()
```
Comment (helto4real) [inline src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs:44]: As comment before. This method can be async. Provide a CancellationToken and use ConnectAsync here using the cancellation token for cancel the potentially long running task connecting. I am sure the Mqtt client has a way to provide a cancellation token to the connect method.

File: src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        _connectionTask?.Wait();
+
+        var retryCount = 1;
+        while ((_mqttClient == null || !_mqttClient.IsConnected) && retryCount <= RetryMaximumSeconds)
+        {
+            _logger.LogDebug("MQTT client is not yet available, retry #{RetryCount}", retryCount);
+            Thread.Sleep(1000);
```
Comment (helto4real) [inline src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs:52]: Using Thread.Sleep in a method that are not async but called from the async code could halt processing and create deadlocks.

**Review by helto4real — CHANGES_REQUESTED** [review]

**Review by helto4real — CHANGES_REQUESTED** [review]
The way of how async/sync code is written will be trouble I think. We should be able to make the GetClientOrThrow async.

File: src/Extensions/NetDaemon.Extensions.MqttEntityManager/MessageSender.cs (hunk 1 above)
Comment (skotl) [reply src/Extensions/NetDaemon.Extensions.MqttEntityManager/MessageSender.cs]: I don't think we need the cancellationtoken in the new approach now, as the managed MQTT client will have its own timeout mechanism - would you agree?

File: src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs (hunk 2 above)
Comment (skotl) [reply src/Extensions/NetDaemon.Extensions.MqttEntityManager/AssuredMqttConnection.cs]: As previous comment

**Review by helto4real — APPROVED** [review]
Looks way better now. Good job.

**Comment by helto4real** [comment]
Closing this, already merged.

**Pull request closed without merging**